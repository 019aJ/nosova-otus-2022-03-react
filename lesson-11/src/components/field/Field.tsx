import { Component } from "react"
import UserService from "../../api/UserService"
import { User } from "../../dto/User"
import UserCard from "../usercard/UserCard"
import styles from "./Field.module.css"
interface FieldProps {
  title: string
  width: number
  height: number
}

interface FieldState {
  selectedCell?: string | null
  isLoading: boolean
  users: User[]
}

const cellMargin = 5

export default class Field extends Component<FieldProps, FieldState> {
  static defaultProps = {
    title: "Пользователи",
    width: 800,
    height: 800,
  }

  constructor(props: FieldProps) {
    super(props)
    this.state = {
      selectedCell: null,
      isLoading: true,
      users: [],
    }
    this.renderChildren = this.renderChildren.bind(this)
  }

  componentDidMount = () => {
    UserService.getAllUsers().then((users) => {
      this.setState({ users, isLoading: false })
    })
  }

  render() {
    const width = this.props.width
    const height = this.props.height
    const cellCount = this.state.isLoading ? 0 : this.state.users.length
    const elemInRow = cellCount ? Math.ceil(Math.sqrt(cellCount)) : 0
    const elemWidth = cellCount ? width / elemInRow - cellMargin : 0
    const elemHeight = cellCount ? height / elemInRow - cellMargin : 0
    return (
      <div
        key="fld"
        className={styles.field}
        style={{
          width: width,
          height: height,
        }}
      >
        {cellCount
          ? this.renderChildren(elemWidth, elemHeight, cellCount)
          : this.props.title}
      </div>
    )
  }

  renderChildren(width: number, height: number, cellCount: number) {
    const cards = []
    for (let index = 0; index < cellCount; index++) {
      cards.push(
        <div key={this.state.users[index].id} className={styles.cell}>
          <UserCard
            user={this.state.users[index]}
            width={width - 2}
            height={height - 2}
          ></UserCard>
        </div>
      )
    }
    return cards
  }
}
