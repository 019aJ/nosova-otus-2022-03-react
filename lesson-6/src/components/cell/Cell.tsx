import { Component } from "react"

interface CellProps {
  title?: string
  width?: string
  height?: string
  id?: string
}

interface CellState {
  title?: string
  width?: string
  height?: string
  id: string
}
export default class Cell extends Component<CellProps, CellState> {
  constructor(props: CellProps) {
    super(props)
    this.state = {
      title: props.title || "0",
      width: props.width || "50px",
      height: props.height || "50px",
      id: props.id || Math.floor(Math.random() * 1000).toString(),
    }
  }
  render() {
    return (
      <div
        id={this.state.id}
        style={{
          backgroundColor: "pink",
          width: this.state.width,
          height: this.state.height,
          border: "solid 1px black",
        }}
      >
        <div style={{ textAlign: "center", position: "relative", top: "30%" }}>
          {this.props.title}
        </div>
      </div>
    )
  }
}
