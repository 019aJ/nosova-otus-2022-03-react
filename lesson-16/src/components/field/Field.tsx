import { initialize } from "../../game/GameAlgorithm"
import { Cell } from "../cell/Cell"
import styles from "./Field.module.css"
import {
  GAME_TITLE,
  FIELD_WIDTH,
  FIELD_HEIGHT,
  FIELD_CELL_COUNT,
} from "../../game/GameDefaults"
import { useDispatch, useSelector } from "react-redux"
interface FieldProps {
  title?: string
  width?: number
  height?: number
  cellCount?: number
}

export const Field: React.FC<FieldProps> = ({
  title,
  width,
  height,
  cellCount,
}) => {
  const renderChildren = (width: number, height: number, cellCount: number) => {
    const values = useSelector<boolean[], boolean[]>((state) => state)
    const dispatch = useDispatch()

    const cards = []
    for (let index = 0; index < cellCount; index++) {
      cards.push(
        <Cell
          key={index.toString()}
          id={index.toString()}
          alive={values[index]}
          width={width}
          height={height}
          onClick={() => {
            dispatch({ type: "MUTATE", payload: { cellIndex: index } })
          }}
        ></Cell>
      )
    }
    return cards
  }

  const elemInRow = cellCount ? Math.ceil(Math.sqrt(cellCount)) : 0
  const elemWidth = cellCount && width ? width / elemInRow : 0
  const elemHeight = cellCount && height ? height / elemInRow : 0
  width = elemInRow > 0 ? elemInRow * elemWidth : width
  height =
    cellCount && elemInRow > 0
      ? Math.ceil(cellCount / elemInRow) * elemHeight
      : height
  return (
    <div key="fld2">
      <div className={styles.header}>{title}</div>
      <div
        key="fld"
        className={styles.field}
        style={{
          width,
          height,
        }}
      >
        {cellCount
          ? renderChildren(elemWidth - 2, elemHeight - 2, cellCount)
          : ""}
      </div>
    </div>
  )
}

Field.defaultProps = {
  title: GAME_TITLE,
  width: FIELD_WIDTH,
  height: FIELD_HEIGHT,
  cellCount: FIELD_CELL_COUNT,
}
