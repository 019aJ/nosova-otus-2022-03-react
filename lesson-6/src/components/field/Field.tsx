import React, { Component } from "react"
import Cell from "../cell/Cell"
import { extractNum, extractUnits } from "../../utils/unitUtils"
interface FieldProps {
  title?: string
  width?: string
  height?: string
  cellCount?: number
}

interface FieldState {
  title: string
  width: string
  height: string
  selectedCell?: string | null
  cellCount: number
}

export default class Field extends Component<FieldProps, FieldState> {
  constructor(props: FieldProps) {
    super(props)
    this.state = {
      title: props.title || "Field",
      width: props.width || "500px",
      height: props.height || "500px",
      selectedCell: null,
      cellCount: props.cellCount || 0,
    }
    this.renderChildren = this.renderChildren.bind(this)
    this.renderRow = this.renderRow.bind(this)
    this.onClick = this.onClick.bind(this)
  }

  onClick = (e: React.MouseEvent<HTMLDivElement>) => {
    let clickedElement = e.target as HTMLElement
    while (!clickedElement.id && clickedElement.parentElement) {
      clickedElement = clickedElement.parentElement
    }
    this.setState({ selectedCell: clickedElement.id })
  }

  render() {
    const width = this.state.width
    const height = this.state.height
    const title = this.state.title
    const cellCount = this.state.cellCount
    if (!cellCount) {
      return <div style={this.ownStyle(width, height)}>{title}</div>
    } else {
      const elemInRow = Math.ceil(Math.sqrt(cellCount))
      const elemWidth = extractNum(width) / elemInRow
      const elemHeight = extractNum(height) / elemInRow
      const unit = extractUnits(width)
      return (
        <div style={this.ownStyle(width, height)} onClick={this.onClick}>
          {this.renderChildren(elemWidth, elemHeight, unit, elemInRow)}
        </div>
      )
    }
  }
  renderCell(id: string, width: number, height: number, unit: string) {
    return (
      <Cell
        key={id}
        title={this.state.selectedCell === id ? id : "click me"}
        width={width - 2 + unit}
        height={height - 2 + unit}
        id={id}
      ></Cell>
    )
  }

  renderRow(
    startIndex: number,
    elemInRow: number,
    width: number,
    height: number,
    unit: string
  ) {
    const rows = []
    const length = this.state.cellCount
    for (let i = startIndex; i < startIndex + elemInRow && i < length; i++) {
      rows.push(
        <div
          style={{
            display: "inline-block",
          }}
        >
          {this.renderCell((i + 1).toString(), width, height, unit)}
        </div>
      )
    }
    return <div style={{ whiteSpace: "nowrap" }}>{rows}</div>
  }

  renderChildren(
    width: number,
    height: number,
    unit: string,
    elemInRow: number
  ) {
    let currentRow = 0
    const rows = []
    for (let index = 0; index < elemInRow; index++) {
      rows.push(this.renderRow(currentRow, elemInRow, width, height, unit))
      currentRow += elemInRow
    }
    return rows
  }

  ownStyle = (width: string, height: string) => {
    return {
      backgroundColor: "yellow",
      width: width,
      height: height,
    }
  }
}


