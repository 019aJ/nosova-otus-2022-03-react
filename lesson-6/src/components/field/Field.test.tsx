import { render } from "@testing-library/react"
import Cell, { CellProps } from "../cell/Cell"
import Field from "../field/Field"

describe("Field snapshot tests", () => {
  it("base snapshot", () => {
    const { container } = render(<Field />)
    expect(container).toMatchSnapshot()
  })

  it("with cells snapshot", () => {
    const { container } = render(<Field cellCount={12} />)
    expect(container).toMatchSnapshot()
  })

  it("some data with sizes snapshot", () => {
    const { container } = render(
      <Field cellCount={12} height="500px" width="300px" />
    )
    expect(container).toMatchSnapshot()
  })
})
