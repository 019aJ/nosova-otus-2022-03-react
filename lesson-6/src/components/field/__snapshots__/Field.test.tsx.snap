// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Field snapshot tests base snapshot 1`] = `
<div>
  <div
    style="background-color: yellow; width: 500px; height: 500px;"
  >
    Field
  </div>
</div>
`;

exports[`Field snapshot tests some data with sizes snapshot 1`] = `
<div>
  <div
    style="background-color: yellow; width: 300px; height: 500px;"
  >
    <div
      style="white-space: nowrap;"
    >
      <div
        style="display: inline-block;"
      >
        <div
          id="1"
          style="background-color: pink; width: 73px; height: 123px; border: 1px solid black;"
        >
          <div
            style="text-align: center; position: relative; top: 30%;"
          >
            click me
          </div>
        </div>
      </div>
      <div
        style="display: inline-block;"
      >
        <div
          id="2"
          style="background-color: pink; width: 73px; height: 123px; border: 1px solid black;"
        >
          <div
            style="text-align: center; position: relative; top: 30%;"
          >
            click me
          </div>
        </div>
      </div>
      <div
        style="display: inline-block;"
      >
        <div
          id="3"
          style="background-color: pink; width: 73px; height: 123px; border: 1px solid black;"
        >
          <div
            style="text-align: center; position: relative; top: 30%;"
          >
            click me
          </div>
        </div>
      </div>
      <div
        style="display: inline-block;"
      >
        <div
          id="4"
          style="background-color: pink; width: 73px; height: 123px; border: 1px solid black;"
        >
          <div
            style="text-align: center; position: relative; top: 30%;"
          >
            click me
          </div>
        </div>
      </div>
    </div>
    <div
      style="white-space: nowrap;"
    >
      <div
        style="display: inline-block;"
      >
        <div
          id="5"
          style="background-color: pink; width: 73px; height: 123px; border: 1px solid black;"
        >
          <div
            style="text-align: center; position: relative; top: 30%;"
          >
            click me
          </div>
        </div>
      </div>
      <div
        style="display: inline-block;"
      >
        <div
          id="6"
          style="background-color: pink; width: 73px; height: 123px; border: 1px solid black;"
        >
          <div
            style="text-align: center; position: relative; top: 30%;"
          >
            click me
          </div>
        </div>
      </div>
      <div
        style="display: inline-block;"
      >
        <div
          id="7"
          style="background-color: pink; width: 73px; height: 123px; border: 1px solid black;"
        >
          <div
            style="text-align: center; position: relative; top: 30%;"
          >
            click me
          </div>
        </div>
      </div>
      <div
        style="display: inline-block;"
      >
        <div
          id="8"
          style="background-color: pink; width: 73px; height: 123px; border: 1px solid black;"
        >
          <div
            style="text-align: center; position: relative; top: 30%;"
          >
            click me
          </div>
        </div>
      </div>
    </div>
    <div
      style="white-space: nowrap;"
    >
      <div
        style="display: inline-block;"
      >
        <div
          id="9"
          style="background-color: pink; width: 73px; height: 123px; border: 1px solid black;"
        >
          <div
            style="text-align: center; position: relative; top: 30%;"
          >
            click me
          </div>
        </div>
      </div>
      <div
        style="display: inline-block;"
      >
        <div
          id="10"
          style="background-color: pink; width: 73px; height: 123px; border: 1px solid black;"
        >
          <div
            style="text-align: center; position: relative; top: 30%;"
          >
            click me
          </div>
        </div>
      </div>
      <div
        style="display: inline-block;"
      >
        <div
          id="11"
          style="background-color: pink; width: 73px; height: 123px; border: 1px solid black;"
        >
          <div
            style="text-align: center; position: relative; top: 30%;"
          >
            click me
          </div>
        </div>
      </div>
      <div
        style="display: inline-block;"
      >
        <div
          id="12"
          style="background-color: pink; width: 73px; height: 123px; border: 1px solid black;"
        >
          <div
            style="text-align: center; position: relative; top: 30%;"
          >
            click me
          </div>
        </div>
      </div>
    </div>
    <div
      style="white-space: nowrap;"
    />
  </div>
</div>
`;

exports[`Field snapshot tests with cells snapshot 1`] = `
<div>
  <div
    style="background-color: yellow; width: 500px; height: 500px;"
  >
    <div
      style="white-space: nowrap;"
    >
      <div
        style="display: inline-block;"
      >
        <div
          id="1"
          style="background-color: pink; width: 123px; height: 123px; border: 1px solid black;"
        >
          <div
            style="text-align: center; position: relative; top: 30%;"
          >
            click me
          </div>
        </div>
      </div>
      <div
        style="display: inline-block;"
      >
        <div
          id="2"
          style="background-color: pink; width: 123px; height: 123px; border: 1px solid black;"
        >
          <div
            style="text-align: center; position: relative; top: 30%;"
          >
            click me
          </div>
        </div>
      </div>
      <div
        style="display: inline-block;"
      >
        <div
          id="3"
          style="background-color: pink; width: 123px; height: 123px; border: 1px solid black;"
        >
          <div
            style="text-align: center; position: relative; top: 30%;"
          >
            click me
          </div>
        </div>
      </div>
      <div
        style="display: inline-block;"
      >
        <div
          id="4"
          style="background-color: pink; width: 123px; height: 123px; border: 1px solid black;"
        >
          <div
            style="text-align: center; position: relative; top: 30%;"
          >
            click me
          </div>
        </div>
      </div>
    </div>
    <div
      style="white-space: nowrap;"
    >
      <div
        style="display: inline-block;"
      >
        <div
          id="5"
          style="background-color: pink; width: 123px; height: 123px; border: 1px solid black;"
        >
          <div
            style="text-align: center; position: relative; top: 30%;"
          >
            click me
          </div>
        </div>
      </div>
      <div
        style="display: inline-block;"
      >
        <div
          id="6"
          style="background-color: pink; width: 123px; height: 123px; border: 1px solid black;"
        >
          <div
            style="text-align: center; position: relative; top: 30%;"
          >
            click me
          </div>
        </div>
      </div>
      <div
        style="display: inline-block;"
      >
        <div
          id="7"
          style="background-color: pink; width: 123px; height: 123px; border: 1px solid black;"
        >
          <div
            style="text-align: center; position: relative; top: 30%;"
          >
            click me
          </div>
        </div>
      </div>
      <div
        style="display: inline-block;"
      >
        <div
          id="8"
          style="background-color: pink; width: 123px; height: 123px; border: 1px solid black;"
        >
          <div
            style="text-align: center; position: relative; top: 30%;"
          >
            click me
          </div>
        </div>
      </div>
    </div>
    <div
      style="white-space: nowrap;"
    >
      <div
        style="display: inline-block;"
      >
        <div
          id="9"
          style="background-color: pink; width: 123px; height: 123px; border: 1px solid black;"
        >
          <div
            style="text-align: center; position: relative; top: 30%;"
          >
            click me
          </div>
        </div>
      </div>
      <div
        style="display: inline-block;"
      >
        <div
          id="10"
          style="background-color: pink; width: 123px; height: 123px; border: 1px solid black;"
        >
          <div
            style="text-align: center; position: relative; top: 30%;"
          >
            click me
          </div>
        </div>
      </div>
      <div
        style="display: inline-block;"
      >
        <div
          id="11"
          style="background-color: pink; width: 123px; height: 123px; border: 1px solid black;"
        >
          <div
            style="text-align: center; position: relative; top: 30%;"
          >
            click me
          </div>
        </div>
      </div>
      <div
        style="display: inline-block;"
      >
        <div
          id="12"
          style="background-color: pink; width: 123px; height: 123px; border: 1px solid black;"
        >
          <div
            style="text-align: center; position: relative; top: 30%;"
          >
            click me
          </div>
        </div>
      </div>
    </div>
    <div
      style="white-space: nowrap;"
    />
  </div>
</div>
`;
